src/redux/services/api.ts

```diff
@@ -34,12 +34,14 @@ export const api = createApi({
   reducerPath: 'api',
   baseQuery: fetchBaseQuery({
     baseUrl: getApiBaseUrl(),
-    prepareHeaders: (headers, { getState }) => {
+    prepareHeaders: async (headers, { getState }) => {
       headers.set('Content-Type', 'application/json');
 
-      const apiToken = (getState() as RootState).tokens.apiToken;
+      const tokens = getApiTokensFromStorage();
```

Comment:
HDS removes API tokens while refreshing either user tokens or api tokens. ( https://helsinkisolutionoffice.atlassian.net/browse/HDS-2141 ). I think in this kind of situation we should wait for the token refresh to complete. Since prepareHeaders can be async, this should be fairly easy to implement. In my implementation it waits for up to 2 seconds to get the API token before throwing an error. This should be enough under normal circumstances for the token refresh to complete.

https://github.com/City-of-Helsinki/pysakoinnin-sahk-asiointi-ui/pull/169/commits/f2dcc6bd52007a35b059b09054387698ca626684#diff-3dabf155febbf2775b0b04a4b659514541a8cba8e2674c4d2b0a8b3bc41f8966R7-R27

then could do something like 

```python
      const tokens = await waitForApiTokens()
```

Reply:
I previously already tried similar approach, and now tried again using the suggestion as a basis, but couldn't get the api tokens to appear for some reason.

Reply:
Yeah I think you can resolve that by having `useApiTokensClient()` present always, for example in the  `App` component. Don't need to use the return value or store it in a variable. This should work:

```
const App = (): React.ReactElement => {
  useApiTokensClient();
  return <WithAuthentication AuthorisedComponent={Authenticated} UnauthorisedComponent={Unauthenticated} />;
};
```

With this the other change related to token fetching, `const tokens = fetch(user);` should no longer be needed in `HandleCallback.tsx` .

Reply:
I tried with this:

```typescript

const RETRY_TIME_MS = 200;
const MAX_RETRY_ATTEMPTS = 10;

const wait = (waitForMs: number) =>
  new Promise<void>(resolve => {
    setTimeout(() => resolve(), waitForMs);
  });

const getApiToken = () => {
  const tokens = getApiTokensFromStorage();
  return tokens ? tokens[String(process.env[`REACT_APP_API_AUDIENCE`])] : undefined;
};

export const waitForApiToken = async () => {
  const retryAttempts = 0;
  while (retryAttempts <= MAX_RETRY_ATTEMPTS) {
    const apiToken = getApiToken();
    if (apiToken) {
      return apiToken;
    }
    await wait(RETRY_TIME_MS);
  }
  throw new Error('Failed to get API token');
};

// Define a service using a base URL and expected endpoints
export const api = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: getApiBaseUrl(),
    prepareHeaders: async (headers, { getState }) => {
      headers.set('Content-Type', 'application/json');
      const apiToken = await waitForApiToken();
      headers.append('authorization', `Bearer ${apiToken}`);
      return headers;
    },
  }),

// rest of the file omitted
```

Reply:
With provided examples I got that part working, thanks! However, removing `const tokens = fetch(user)` is still not possible. Removing it causes weird unexpected behavior where login fails because user apparently is already logged in.

Reply:
Aha, yes, I see now. WithAuthentication does not re-render automatically if authentication changes. There are a few ways to go about it, but the most simple is to use `useOidcClientTracking()` which will force a re-render when user is updated, removed and being renewed:

```typescript
const App = (): React.ReactElement => {
  useOidcClientTracking();
  useApiTokensClient();
  return <WithAuthentication AuthorisedComponent={Authenticated} UnauthorisedComponent={Unauthenticated} />;
};
```

useApiTokensClient() does not force any re-renders, but that is fine since the api token is fetched manually on demand in api.ts.

Reply:
OK, lets proceed with this approach then.